### PR TITLE
ENH: stats: Add the function odds_ratio.

### LINKED ISSUE #11131: DOC: stats.fisher_exact does not match R functionality for `oddsratio` calculation
This a known issue, mentioned in [gh-1483](https://github.com/scipy/scipy/issues/1483#issuecomment-17030569) and reflected by [the docstring for `stats.fisher_exact`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.fisher_exact.html):
> The calculated odds ratio is different from the one R uses. This scipy implementation returns the (more common) “unconditional Maximum Likelihood Estimate”, while R uses the “conditional Maximum Likelihood Estimate”.

But despite that documentation, this discrepancy nevertheless still causes confusion.  There is also some strong disagreement (e.g. [this conversation on CrossValidated](https://stats.stackexchange.com/questions/288641/odds-ratio-mle-calculation-unconditional-vs-conditional)) with SciPy's choice to use an "unconditional" MLE (here, the sample odds ratio) in the first place.  I am not a statistician, but my research (some rough notes [here](https://hackmd.io/AAmCAcryRWu2BKGZfc9wew)) leads me to agree.  While the sample odds ratio is arguably "more common" throughout statistics, it seems that the literature for Fisher's exact test only supports the use of CMLE.

Are there any plans to match R functionality for `fisher_exact`?  This was recently done for `hmean` with gh-10899 by adding an optional parameter.  If not, I propose to correct and expand the docstring to better explain the issue, and would welcome any expert input to help me with my understanding of the issue.

### PR DESCRIPTION
**Pull request by WarrenWeckesser**
The new function `scipy.stats.contingency.odds_ratio` computes
the odds ratio and p-value for a 2x2 contingency table.  An option
allows the user to select either the sample odds ratio or the
conditional maximum likelihood estimate of the odds ratio.  The
returned object provides a method for computing the confidence
interval of the odds ratio.

Closes gh-11131.

### CODE REVIEW

**Comment by WarrenWeckesser** [comment]
This PR replaces gh-12288. This PR creates a new function in the `scipy.stats.contingency` namespace specifically for the odds ratio statistic, similar to how the proposed function `relative_risk` in gh-13048 creates a function for another widely used statistic for 2x2 contingency tables, the relative risk.

There is code in the file `_odds_ratio.py` that implements a private version of Fisher's noncentral hypergeometric distribution.  Some of that code will be able to use the implementation in gh-13330 instead, but we don't necessarily have to block this PR waiting for gh-13330.  If it turns out that gh-13330 is merged first, I can update this PR to use it, but if this PR goes in first,  the private implementation can be updated after gh-13330 is merged.

File: scipy/stats/stats.py
Comment (sethtroisi) [inline scipy/stats/stats.py:4009]: nit: different number of spaces between ". This" and ".  To compute"

File: scipy/stats/tests/generate_fisher_exact_results_from_r.R
Comment (sethtroisi) [inline scipy/stats/tests/generate_fisher_exact_results_from_r.R:1]: This is very nice.
Maybe add a small comment on how to run e.g. `$ Rscript fisher_exact_results_from_r.py`

Comment (sethtroisi) [inline scipy/stats/tests/generate_fisher_exact_results_from_r.R:1]: Copyright header?

**Review by sethtroisi — COMMENTED** [review]
Small comments, I hope to dive into `_odds_ratio.py` tomorrow.

File: scipy/stats/tests/generate_fisher_exact_results_from_r.R
Comment (WarrenWeckesser) [reply scipy/stats/tests/generate_fisher_exact_results_from_r.R:1]: I wrote the script, so it is under the same license as all of the developer contributions to SciPy.  We generally don't add copyright notices to our code.

**Comment by mdhaber** [comment]
> Some of that code will be able to use the implementation in gh-13330 instead, but we don't necessarily have to block this PR waiting for gh-13330.

@WarrenWeckesser you proposed the addition of the two noncentral hypergeometric distributions. I implemented them in gh-13330 as we discussed. Rather than asking others to review the private code, then re-review when you swap things out, would it be more efficient for you to review gh-13330 first?

**Comment by WarrenWeckesser** [comment]
@mdhaber, yes, that makes sense.  I'll mark this as draft for now.

**Comment by WarrenWeckesser** [comment]
I have updated this PR to use the new `nchypergeom_fisher` distribution.

File: scipy/stats/__init__.py
Comment (mdhaber) [inline scipy/stats/__init__.py:446]: The plan is to remove this and get the documentation for `OddsRatioResult` to render otherwise, right?

File: scipy/stats/_odds_ratio.py
Comment (mdhaber) [inline scipy/stats/_odds_ratio.py:338]: I don't think we want this to be a bulleted list, right?
See e.g. [`linprog` documentation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.linprog.html) of the return object, an `OptimizeResult`:
image

File: scipy/stats/_odds_ratio.py
Comment (mdhaber) [inline scipy/stats/_odds_ratio.py:340]: As much as I'd like keyword names to appear as teletype everywhere (without the gray background, though...), numpydoc says variables should have single backticks, right?

File: scipy/stats/__init__.py
Comment (WarrenWeckesser) [reply scipy/stats/__init__.py:446]: That is the goal, but so far I haven't been able to get the online docs for `OddsRatioResult` to render without having something like this in the `stats` module docstring in `__init__.py`.

File: scipy/stats/_odds_ratio.py
Comment (WarrenWeckesser) [reply scipy/stats/_odds_ratio.py]: Good idea, done.

File: scipy/stats/_odds_ratio.py
Comment (WarrenWeckesser) [reply scipy/stats/_odds_ratio.py]: Done.

File: scipy/stats/stats.py
Comment (WarrenWeckesser) [reply scipy/stats/stats.py:4211]: Done.

File: scipy/stats/tests/generate_fisher_exact_results_from_r.R
Comment (WarrenWeckesser) [reply scipy/stats/tests/generate_fisher_exact_results_from_r.R:1]: Done.

File: scipy/stats/tests/test_odds_ratio.py
Comment (mdhaber) [inline scipy/stats/tests/test_odds_ratio.py:15]: They are very precise - 15 digits are included. Do you mean they are not very accurate? How do we know that the R results are inaccurate and `odds_ratio` is?

File: scipy/stats/tests/test_odds_ratio.py
Comment (mdhaber) [inline scipy/stats/tests/test_odds_ratio.py:9]: Input validation tests?

File: scipy/stats/stats.py
Comment (mdhaber) [reply scipy/stats/stats.py:4211]: Fortunately Sphinx renders these as single spaces : )

File: scipy/stats/stats.py
Comment (mdhaber) [inline scipy/stats/stats.py:4210]: Is it ever referred to as the "sample" MLE? I only ask because the PR uses sample/conditional instead of "unconditional"/conditional elsewhere.

File: scipy/stats/_odds_ratio.py
Comment (mdhaber) [inline scipy/stats/_odds_ratio.py:162]: LGTM based on what I see in Wikipedia.

**Review by mdhaber — CHANGES_REQUESTED** [review]
So far I've reviewed documentation, tests, and sample odds ratio calculations (odds ratio, p value, and CI). Calculations look good; biggest comments are that sample odds ratio calculations and input validation need tests.

I see the reference for the conditional odds ratio CI. Is there one particular reference that you'd recommend I follow for the conditional odds ratio itself and for the fact that the `fisher_exact` p-value should be the same as this p-value?

File: scipy/stats/tests/test_odds_ratio.py
Comment (WarrenWeckesser) [reply scipy/stats/tests/test_odds_ratio.py:15]: Yes, that should be "... not very accurate."  The comment is based on lots of scratch work with mpmath, and on verifying the odds ratio with tests such as the self-check a few lines down from here.

I'm adding some additional consistency checks for the odds ratio confidence interval, and I'll add the odds ratio calculations to [`mpsci`](https://github.com/WarrenWeckesser/mpsci) so an mpmath-based version of the calculation is available.

File: scipy/stats/__init__.py
Comment (mdhaber) [reply scipy/stats/__init__.py:446]: Unresolving this based on the progress in gh-13048

**Comment by WarrenWeckesser** [comment]
Here are some notes about relevant references, for anyone looking to learn more about inference with the odds ratio.

* The following text covers the calculation in detail, in section 4.1.2.  Specifically, equation (4.7) is the equation that must be solved for the CMLE.

    - H. Sahai and A. Khurshid (1996), Statistics in Epidemiology: Methods, Techniques, and Applications, CRC Press LLC, Boca Raton, Florida.

* This publication is often cited when the CMLE is disussed:

    - Breslow NE, Day NE. Statistical methods in cancer research. Volume I - The analysis of case-control studies. IARC Sci Publ. 1980;(32):5-338. PMID: 7216345.

  The PDF is available at

    https://publications.iarc.fr/Book-And-Report-Series/Iarc-Scientific-Publications/Statistical-Methods-In-Cancer-Research-Volume-I-The-Analysis-Of-Case-Control-Studies-1980

  The odds ratio and the CMLE are discussed in section 4.2, beginning on page 124.  The defining equation for the CMLE is (4.5).

* Inference for the odds ratio is discussed in section 10.3 of the text

    - M. Hollander, D. A. Wolfe and  E. Chicken (2014), Nonparametric Statistical Methods (third edition), John Wiley & Sons, Inc., Hoboken, NJ, USA.

  The CMLE is covered in comments 14, 15 and 16  in the  "Comments" subsection on pages 519-520.  Note that there is a typographical error in comment 16: the reference to equation (10.38) should be (10.58).

File: scipy/stats/_odds_ratio.py
Comment (mdhaber) [inline scipy/stats/_odds_ratio.py:59]: Could you comment on why `xtol` is overridden?

File: scipy/stats/_odds_ratio.py
Comment (mdhaber) [inline scipy/stats/_odds_ratio.py:44]: Looks to me like it will do that.

File: scipy/stats/_odds_ratio.py
Comment (mdhaber) [inline scipy/stats/_odds_ratio.py:70]: Converting between different conventions is annoying, so I haven't. Assuming the Breslow and Day reference equation 2 is an equivalent formulation of Fisher's NCHG distribution with these arguments, this looks right.

Do you want me to check it carefully, or shall I assume that you did and that it is unlikely that the numbers this function produces would match R otherwise?

File: scipy/stats/_odds_ratio.py
Comment (mdhaber) [inline scipy/stats/_odds_ratio.py:107]: The `x-1` here is because `cdf` includes the probability mass at `x` whereas `sf` does not, right?

File: scipy/stats/_odds_ratio.py
Comment (mdhaber) [inline scipy/stats/_odds_ratio.py:117]: The support of the NCHG distribution is independent of the NC parameter, so this is just a way of getting the support without specifying an (arbitrary) NC parameter, right?

File: scipy/stats/_odds_ratio.py
Comment (mdhaber) [inline scipy/stats/_odds_ratio.py:123]: The only way 0 could be the mean of the distribution is if the noncentrality parameter is 0 (so the random variable is identically 0). Otherwise, some of the probability mass would be greater than 0, and therefore the mean would not be 0. Correct?

File: scipy/stats/tests/test_odds_ratio.py
Comment (mdhaber) [inline scipy/stats/tests/test_odds_ratio.py:42]: Should we fix that?

File: scipy/stats/_odds_ratio.py
Comment (WarrenWeckesser) [reply scipy/stats/_odds_ratio.py:395]: No.  I'll just remove that sentence.

File: scipy/stats/_odds_ratio.py
Comment (WarrenWeckesser) [reply scipy/stats/_odds_ratio.py:107]: Yes.  For a discrete distribution, `sf(k, ...)` sums the PMF from k+1 to the right end of the support.

File: scipy/stats/_odds_ratio.py
Comment (WarrenWeckesser) [reply scipy/stats/_odds_ratio.py:117]: Yes.  The support of Fisher's NCHG does not depend on the noncentrality.

File: scipy/stats/_odds_ratio.py
Comment (WarrenWeckesser) [reply scipy/stats/_odds_ratio.py:123]: Yup.

File: scipy/stats/tests/test_odds_ratio.py
Comment (WarrenWeckesser) [reply scipy/stats/tests/test_odds_ratio.py:42]: I don't think we have to, and I'm not sure we should.  Technically, if the odds ratio is 0 or inf, the support collapses to a single value.  So would we also modify the calculation of the support of the nchg distributions in these edge cases?  I think we can defer such a discussion to another time.

File: scipy/stats/_odds_ratio.py
Comment (WarrenWeckesser) [reply scipy/stats/_odds_ratio.py]: The link is above, so I don't think we need to repeat that very long name again.

File: scipy/stats/_odds_ratio.py
Comment (WarrenWeckesser) [reply scipy/stats/_odds_ratio.py:452]: That was for a scalar argument to `binomtest`.  Here we require an integer array.

File: scipy/stats/_odds_ratio.py
Comment (WarrenWeckesser) [reply scipy/stats/_odds_ratio.py:59]: Initially I suspect I used that value because I've gotten used to other solvers in `scipy.minimize` having fairly large relative tolerances, and I forgot that `brentq` defaults to 2e-12.

In a recent commit, I added some additional tests for the case `kind='sample'`.  Now if I use the default `xtol` in `brentq`, four of those tests fail, and they pass with `xtol=1e-13`.

File: scipy/stats/_odds_ratio.py
Comment (WarrenWeckesser) [reply scipy/stats/_odds_ratio.py]: This is probably not quite what you are looking for, but I updated the code to use parameter names that are consistent with the names used in `stats.hypergeom` and `stats.nchypergeom_fisher`.

For a contingency table [[a, b], [c, d]], Breslow and Day's notation is defined in their "equation" (4.1).  The correspondence to the SciPy parameters is
```
SciPy    Breslow & Day     Meaning in terms of 2x2 contingency table
  k           a                    table[0, 0]
  M           N                    sum(table)
  n           n1                   table[0].sum() 
  N           m1                   table[: 0].sum()     
```
In Breslow and Day's notation, pr(a | n1, n0, m1, m0, Ψ) is the PMF for the noncentral hypergeometric distribution; that parameterization is redundant, since n0 + n1 = m0 + m1.

File: scipy/stats/__init__.py
Comment (WarrenWeckesser) [reply scipy/stats/__init__.py]: The documentation for `OddsRatioResult` is now generated the same as the `RelativeRiskResult` and `BinomTestResult`.

File: scipy/stats/_odds_ratio.py
Comment (WarrenWeckesser) [reply scipy/stats/_odds_ratio.py]: I'd like to keep the description of the essential results of `odds_ratio` in its docstring.  "A click away" is not as nice as "no click", and only applies to those reading the docs on the web.  Someone working in an ipython session would have to figure out how to import `OddsRatioResult` before they could read the class's docstring.

For completeness, I copied the details for the `odds_ratio` and `pvalue` attributes to here from the `odds_ratio` docstring.  I know that violates DRY, but I think keeping those details in the `odds_ratio` docstring provides a better user experience.

(Now that we're using these classes in more functions, we probably need to create some style guidelines for issues like this.)

File: scipy/stats/_odds_ratio.py
Comment (WarrenWeckesser) [reply scipy/stats/_odds_ratio.py]: This and the other backtick issues should be fixed now.

File: scipy/stats/_odds_ratio.py
Comment (WarrenWeckesser) [reply scipy/stats/_odds_ratio.py:433]: I updated the example in the docstring.

File: scipy/stats/tests/test_odds_ratio.py
Comment (WarrenWeckesser) [reply scipy/stats/tests/test_odds_ratio.py:9]: I added some input validation tests and tests of the odds ratio and its confidence interval for `kind='sample'`.

File: scipy/stats/stats.py
Comment (WarrenWeckesser) [reply scipy/stats/stats.py:4214]: Yes, good idea.  Presumably this refers to `fisher.test`. I'll update the docstring.

File: scipy/stats/_odds_ratio.py
Comment (mdhaber) [reply scipy/stats/_odds_ratio.py]: OK. I'll let you handle the style guidelines. I really think stats would be a good candidate for Google Season of Docs but I can't manage that right now.

File: scipy/stats/_odds_ratio.py
Comment (mdhaber) [reply scipy/stats/_odds_ratio.py]: OK. I suppose I didn't notice it on the previous line, so it must not be terribly important.

File: scipy/stats/_odds_ratio.py
Comment (mdhaber) [reply scipy/stats/_odds_ratio.py:433]: OK. I don't mind suggesting an interpretation of results as long as it's done carefully, but I can see why we would avoid it.

File: scipy/stats/_odds_ratio.py
Comment (mdhaber) [reply scipy/stats/_odds_ratio.py:460]: This is more of a question about that function - why did you write it if `np.issubdtype` does a good check? 
If I remember correctly, it will also check that the value is scalar and bounded on one side?

File: scipy/stats/_odds_ratio.py
Comment (mdhaber) [reply scipy/stats/_odds_ratio.py:59]: Good enough for me.

File: scipy/stats/_odds_ratio.py
Comment (mdhaber) [reply scipy/stats/_odds_ratio.py]: K. Might be worth a comment as it might look like an oversight.

File: scipy/stats/tests/test_odds_ratio.py
Comment (mdhaber) [reply scipy/stats/tests/test_odds_ratio.py:42]: OK.

File: scipy/stats/tests/test_odds_ratio.py
Comment (mdhaber) [reply scipy/stats/tests/test_odds_ratio.py:9]: : )
Would you agree that they are a llittle less thorough than, say, the tests of the conditional odds ratio?
Do you not think they're important?

**Review by mdhaber — COMMENTED** [review]
Most responses look good. Still need to document alternatives. I would suggest more thorough tests of input validation and sample odds ratio, but if you really think they're unnecessary, let me know. I still need to check the changes to the notation, but otherwise, this is close.

File: scipy/stats/contingency.py
Comment (mdhaber) [inline scipy/stats/contingency.py]: Was removing this from the list intentional?

File: scipy/stats/_odds_ratio.py
Comment (mdhaber) [inline scipy/stats/_odds_ratio.py:76]: You were right, this is exactly not what I would have done. But I compared these against the reference, and this parameterization seems equivalent.

**Comment by mdhaber** [comment]
Before merge:

[ ] - To help future reviewers, add comments to code indicating where to look for the relevant information
[ ] - Comment in code on use of `hypergeom.support` instead of `nchypergeom_fisher.support`
[ ] - Document `alternatives`
[ ] - More thorough unit test for input validation and sample odds ratio calculations
[ ] - Address removal of `association` from `contingency.py`

Conclusion to mailing list discussion about whether this belongs in `scipy.stats` or `scipy.stats.contingency` would be nice, but I would not let that hold this up. This (and other things) can be moved out of `contingency` before release if that's where the conclusion goes.

File: scipy/stats/contingency.py
Comment (WarrenWeckesser) [reply scipy/stats/contingency.py]: Yes, because it is already listed four lines above.

Comment (mdhaber) [reply scipy/stats/contingency.py]: Oops, OK.

File: scipy/stats/_odds_ratio.py
Comment (mdhaber) [inline scipy/stats/_odds_ratio.py:322]:
```suggestion
    Compute the odds ratio for a 2x2 contingency table.
    
    Also tests the null hypothesis that the true odds ratio is 1.
```

File: scipy/stats/_odds_ratio.py
Comment (mdhaber) [inline scipy/stats/_odds_ratio.py:338]:
```suggestion
        * 'two-sided': the true odds ratio is not equal to 1
        * 'less': the true odds ratio is less than 1
        * 'greater': the true odds ratio is greater than 1
```

**Review by mdhaber — COMMENTED** [review]
This was really close. Shall we finish it? I think the remaining request was just documenting these alternatives and for more specific comments to be added to the code to point readers to the appropriate reference + equation.

File: scipy/stats/_odds_ratio.py
Comment (WarrenWeckesser) [reply scipy/stats/_odds_ratio.py:317]: Added in the latest commit.

File: scipy/stats/_odds_ratio.py
Comment (WarrenWeckesser) [reply scipy/stats/_odds_ratio.py]: Added in the latest commit.

**Comment by mdhaber** [comment]
I'm re-running CircleCI. Looked like a temporary glitch.

**Comment by mdhaber** [comment]
Re-ran CircleCI; same failure. Not seeing it in other PRs. But I can't see how the latest two commits would have cause this.

Aside: something I thought about when we were discussion the alternatives - from a UI perspective, how should we think about the relationship between this function and `fisher_exact`? Does this function essentially make `fisher_exact` obsolete but retain it for backwards compatibility (and its more common name)?

File: scipy/stats/_odds_ratio.py
Comment (mdhaber) [inline scipy/stats/_odds_ratio.py:417]:
```suggestion
    >>> res = odds_ratio([[7, 15], [58, 472]])
```

File: scipy/stats/tests/test_odds_ratio.py
Comment (mdhaber) [reply scipy/stats/tests/test_odds_ratio.py:9]: I added more substantial tests (e.g. confidence interval other than the default, one-sided confidence intervals).

**Review by mdhaber — APPROVED** [review]

File: scipy/stats/_odds_ratio.py
Comment (mdhaber) [inline scipy/stats/_odds_ratio.py:3]:
```suggestion
```

File: scipy/stats/_odds_ratio.py
Comment (mdhaber) [inline scipy/stats/_odds_ratio.py:332]: If this stays in `stats.contingency`, we could consider calling this `observed` to match `chi2_contingency`, `expected_freq`, and `association`.

The documentation explains how the table is laid out, so I don't see a need for the four separate arguments of `relative_risk`. 

If this goes in `stats`, I would keep `table` to match `fisher_exact`, `barnard_exact`, and `boschloo_exact`. 

On the other hand, I can see leaving it as `table` either way and using [`_rename_parameter`](https://github.com/scipy/scipy/blob/6988f976a8526c54c57e0a6d349642f0dc8e05f7/scipy/_lib/_util.py#L618) to make the `stats.contingency` functions consistent with `stats`.

File: scipy/stats/_odds_ratio.py
Comment (mdhaber) [inline scipy/stats/_odds_ratio.py:333]: There is not much precedent for `kind` in `scipy.stats`, but `method` is not appropriate IMO since it's more a matter of choice/definition rather than algorithm, and I'd avoid the keyword `type`.

**Review by mdhaber — APPROVED** [review]

**Comment by mdhaber** [comment]
No responses to [email](https://mail.python.org/archives/list/scipy-dev@python.org/thread/GZL6QNHO4GNLVPAOSWUATM62JPP6RJOB/), but this seems fundamental, was requested by a user, and at least two maintainers have wanted to have this function, so let's merge it. For now, this was trimmed down to the bare minimum (statistic and confidence interval), but if we want to change parameter names, add output, etc., that can all be done in a backward compatible way.

**Pull request merged**
